gateways/firstdata/firstdata.go

```diff
@@ -0,0 +1,257 @@
+package firstdata
+
+import (
+	"bytes"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/BoltApp/sleet"
+	"github.com/BoltApp/sleet/common"
+)
+
+const (
+	authPath = "/payments"
+)
+
+type FirstdataClient struct {
+	host            string
+	apiKey          string
+	apiSecret       string
+	clientRequestID string
+	httpClient      *http.Client
+}
+
+func NewClient(env common.Environment, apiKey, apiSecret string) *FirstdataClient {
+	return NewWithHttpClient(env, apiKey, apiSecret, common.DefaultHttpClient())
+}
+
+func NewWithHttpClient(env common.Environment, apiKey, apiSecret string, httpClient *http.Client) *FirstdataClient {
+	return &FirstdataClient{
+		host:       firstdataHost(env),
+		apiKey:     apiKey,
+		httpClient: httpClient,
+	}
+}
+
+func (client *FirstdataClient) Authorize(request *sleet.AuthorizationRequest) (*sleet.AuthorizationResponse, error) {
+	firstdataAuthRequest, err := buildAuthRequest(request)
+
+	clientRequestId := request.ClientTransactionReference
+
+	if err != nil {
+		return nil, err
+	}
+
+	url := "https://" + client.host + authPath
+
+	firstdataResponse, err := client.sendRequest(*clientRequestId, url, *firstdataAuthRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	success := false
+
+	if firstdataResponse.TransactionStatus == "APPROVED" || firstdataResponse.TransactionStatus == "WAITING" {
+		success = true
+	}
+
+	if firstdataResponse.Error != nil {
+		response := sleet.AuthorizationResponse{Success: false, ErrorCode: firstdataResponse.Error.Code}
+		return &response, nil
+	}
+
+	avsRaw, err := json.Marshal(firstdataResponse.Processor.AVSResponse)
```

Comment:
This should already be marshalled right? Why not just call AvsRawString = string(firstdataResponse.Processor.AVSResponse?)

Reply:
AVSResponse is a struct. string(AVSResponse) would throw a cannot convert error. I could use fmt.Sprintf("%v",AVSResponse) if you would prefer that.

Reply:
Wait I guess I meant, when you unmarshalled the response into the firstdataResponse, AVS was also unmarshalled right?

Reply:
Yes, it was UnMarshalled from a Json string into a Go struct, at the end of the send function, which then returns that Go struct. It can't (as far as i know, i tried your suggestion just now and it was an error) be automatically converted back into a string without marshaling it back to a json string with json.Marshal().

```
type ProcessorData struct {
        ...
	AVSResponse             AVSResponse `json:"avsResponse"`
}

type AVSResponse struct {
	StreetMatch   string `json:"streetMatch"`  
	PostCodeMatch string `json:"postCodeMatch"` 
}

```

Reply:
So at the end `avsRaw` will be something like `{"streetMatch": "Y", "postCodeMatch": "Y"}`? I think it might be cleaner to actually just do something like fmt.Sprintf("%s:%s", avsResponse.StreetMatch, avsResponse.PostCodeMatch) so we don't put the "streetMatch" and "postCodeMatch" strings in there